### PR TITLE
chore: format rows worker

### PR DESCRIPTION
**Pull request by rephus**
Closes: #4390

### Description:

Before the api was locked when doing `formatRows` now it is not locked because it runs on a different thread

### CODE REVIEW

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 1]
```diff
@@ -587,12 +617,16 @@ export class ProjectService {
...
         const { rows } = await adapter.runQuery(query);
 
-        const formattedRows = formatRows(
-            rows,
+        const itemMap = getItemMap(
             explore,
             metricQuery.additionalMetrics,
             metricQuery.tableCalculations,
         );
+        const formattedRows = await wrapSentryTransaction(
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:625]: we can remove the `wrapSentryTransaction` usage here (or not), I added it for testing

File: packages/backend/src/services/ProjectService/formatRows.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+
+const { workerData, parentPort } = require("worker_threads");
+
+const { formatItemValue } = require("@lightdash/common");
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/formatRows.js:4]: I'm not sure if there is a better way to do the import here... unfortunatelly this method is too complex to just copy it here

Comment (rephus) [reply packages/backend/src/services/ProjectService/formatRows.js:4]: btw, worker_thread only allows simple values, we can't pass functions

File: packages/backend/src/utils.ts [hunk 3]
```diff
@@ -27,3 +30,41 @@ export const isDbPinnedChart = (
     data: DbPinnedChart | DbPinnedDashboard,
 ): data is DbPinnedChart =>
     'saved_chart_uuid' in data && !!data.saved_chart_uuid;
+
+export const wrapSentryTransaction = async (
```
Comment (rephus) [inline packages/backend/src/utils.ts:34]: Wrap method to add spans to sentry 

https://lightdash.sentry.io/performance/lightdash:0920559162d24063a216e9cf13cc4528/?project=5959292&query=%2Fapi%2Fv1%2Fprojects%2F963e0810-ecac-47dc-b170-444c84ea0601%2Fexplores%2F%3AexploreId%2FdownloadCsv&referrer=performance-transaction-summary&statsPeriod=24h&transaction=POST+%2Fapi%2Fv1%2Fprojects%2F963e0810-ecac-47dc-b170-444c84ea0601%2Fexplores%2F%3AexploreId%2FdownloadCsv&unselectedSeries=p100%28%29

image

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 4]
```diff
@@ -587,12 +617,16 @@ export class ProjectService {
...
+            'formatted rows',
+            {},
+            async () => formatRowsWorker(rows, itemMap),
```
Comment (rephus) [inline packages/backend/src/services/ProjectService/ProjectService.ts:628]: I think we should only run this on the worker when rows.length > 500,

File: packages/backend/src/utils.ts [hunk 5]
```diff
@@ -27,3 +30,41 @@ export const isDbPinnedChart = (
...
+    name: string,
+    context: CustomSamplingContext,
+    funct: () => Promise<any>,
```
Comment (ZeRego) [inline packages/backend/src/utils.ts:37]: We can improve the function typing to return the same type as `funct`
```
export const wrapSentryTransaction = async <T>(
    name: string,
    context: CustomSamplingContext,
    funct: () => Promise<T>,
): Promise<T> =>  {
```

File: packages/backend/src/utils.ts [hunk 6]
```diff
@@ -27,3 +30,41 @@ export const isDbPinnedChart = (
...
+            data: context,
+        });
+    try {
+        return await funct();
+    } catch (error) {
+        Logger.error(
+            `Error in wrapped sentry transaction ${transaction?.spanId} "${name}": ${error}`,
+        );
+        Sentry.captureException(error);
+        return null;
```
Comment (ZeRego) [inline packages/backend/src/utils.ts:61]: I would expect to throw the same error instead of null

File: packages/backend/src/utils.ts (hunk 3 above)
Comment (ZeRego) [reply packages/backend/src/utils.ts:34]: This wrapper is really useful 👨‍🍳 💋

**Comment by rephus** [comment]
Added context about the query 

image
Screenshot from 2023-02-13 13-47-03

File: packages/backend/src/services/ProjectService/formatRows.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+/* tslint:disable */
+/* eslint-disable */
+const { workerData, parentPort } = require('worker_threads');
+
+const { formatItemValue } = require('@lightdash/common');
+
+function formatRows(rows, itemMap) {
+    const formattedRows = rows.map((row) =>
```
Comment (ZeRego) [inline packages/backend/src/services/ProjectService/formatRows.js:8]: Couldn't we just import the `formatRows` from common? 🤔 
We can update that function to receive the itemMap as we did here.

**Review by ZeRego — APPROVED** [review]
Looks good 👍 
Just had a suggestion to reuse the existing util function.

File: packages/backend/src/services/ProjectService/formatRows.js (hunk 7 above)
Comment (rephus) [reply packages/backend/src/services/ProjectService/formatRows.js]: Nice suggestion, done

**Pull request merged**